pkg/controller/storagecluster/update.go

```diff
@@ -332,6 +337,70 @@ func (c *Controller) getKVDBNodeAvailability(cluster *corev1.StorageCluster) (in
 
 }
 
+func (c *Controller) parallelUpgradeNodesList(cluster *corev1.StorageCluster, pods []*v1.Pod) []*v1.Pod {
+	if len(pods) == 0 {
+		return pods
+	}
+
+	if storagePodsEnabled(cluster) {
+		storageNodeList, err := c.Driver.GetStorageNodes(cluster)
+		if err != nil {
+			logrus.Warnf("Cannot upgrade nodes in parallel, failed to get storage nodes: %v", err)
+			return pods
+		}
+		if !util.ClusterSupportsParallelUpgrade(storageNodeList) {
+			logrus.Infof("Skipping parallel upgrade as not all nodes are running minimum required version : 3.1.2")
+			return pods
+		}
+		nodeNameToIdMap := make(map[string]string)
+		nodesToUpgrade := make([]string, 0)
+		nodeIdToPodMap := make(map[string]*v1.Pod)
+		for _, storageNode := range storageNodeList {
+			nodeNameToIdMap[storageNode.SchedulerNodeName] = storageNode.Id
+		}
+
+		for _, pod := range pods {
+			if nodeId, ok := nodeNameToIdMap[pod.Spec.NodeName]; ok {
+				nodesToUpgrade = append(nodesToUpgrade, nodeId)
+				nodeIdToPodMap[nodeId] = pod
+			}
+
+		}
+
+		c.sdkConn, err = util.GetPortworxConn(c.sdkConn, c.client, cluster.Namespace)
+		if err != nil {
+			logrus.Warnf("Cannot upgrade nodes in parallel, failed to get Portworx connection: %v", err)
+			return pods
+		}
+		nodeClient := storageapi.NewOpenStorageNodeClient(c.sdkConn)
+		ctx, err := util.SetupContextWithToken(context.Background(), cluster, c.client, false)
+		if err != nil {
+			logrus.Warnf("Cannot upgrade nodes in parallel, failed to setup context with token: %v", err)
+			return pods
+		}
+		nodeFilterResponse, err := nodeClient.FilterNonOverlappingNodes(
```

Comment:
Let's not expose the portworx grpc client here as we generally put all PX specific logic under drivers/storage/portworx. 
Let's add a new wrapper method to the [Driver interface](https://github.com/libopenstorage/operator/blob/master/drivers/storage/storage.go#L26) and then implement the method in portworx.go.

This will also help us mock the driver interface easily in the UTs.